WalletWasabi.Tests/UnitTests/SmartBlockProviderTests.cs

```diff
@@ -21,31 +23,39 @@ public async void TestAsync()
 				[uint256.One] = Block.CreateBlock(Network.Main)
 			};
 
-			var source = new MockProvider();
-			source.OnGetBlockAsync = async (hash, cancel) =>
-			{
-				await Task.Delay(TimeSpan.FromSeconds(0.5));
-				return blocks[hash];
-			};
+			var blockProvider = new TestBlockProvider(blocks);
```

Comment:
The MockXXXX pattern that we are using here, and in other places, is more flexible than this one because it allows us to test many different scenarios, for example, imagine I want to test how the `SmartBlockProvider` behaves when it receives a RPCException, when the cache block is not found, when the task times out, etc.

Reply:
I see your point. 

I attempt to implement only what is necessary because the more complicate the test code is the more you risk bugs in actual test code. It's true that the original implementation can exercise the class much more in depth but there are no such tests. So is the abstraction actually needed? 

Often what I encounter is that you have some abstraction in place and you expect somebody just to use it but then somebody comes and he actually needs a little bit different and she does not use your abstraction as expected. That's why I find it tempting not to "predict" what others may need. But it's a bit philosophical answer.

Reply:
Both approach is good. Fight it out! If it's a tie, then we'll merge whatever version is here after the rest of the (also minor, but uncontroversial) issues are resolved.